pkg/kubelet/userns/userns_manager.go

```diff
@@ -379,20 +380,33 @@ func (m *UsernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err err
 }
 
 // GetOrCreateUserNamespaceMappings returns the configuration for the sandbox user namespace
-func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod, runtimeHandler string) (*runtimeapi.UserNamespace, error) {
+	featureEnabled := utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport)
+
+	if pod == nil || pod.Spec.HostUsers == nil || *pod.Spec.HostUsers {
+		// if the feature is enabled, specify to use the node mode...
+		if featureEnabled {
+			return &runtimeapi.UserNamespace{
+				Mode: runtimeapi.NamespaceMode_NODE,
+			}, nil
+		}
+		// ...otherwise don't even specify it
+		return nil, nil
+	}
+
+	// if the pod requested a user namespace and the runtime doesn't support user namespaces then return an error,
+	// even when the feature gate is disabled.
+	if handlerSupportsUserns, err := m.kl.HandlerSupportsUserNamespaces(runtimeHandler); err != nil {
+		return nil, err
+	} else if !handlerSupportsUserns {
+		return nil, fmt.Errorf("the handler %q does not support user namespaces", runtimeHandler)
+	} else if !featureEnabled {
```

Comment:
As I mentioned in the outdated comment, this doesn't reject the pod if the feature gate is disabled. This just ignores the field.

This diff does the trick for me:
```go
diff --git pkg/kubelet/userns/userns_manager.go pkg/kubelet/userns/userns_manager.go
index bfc18e08e22..67d1261e9ed 100644
--- pkg/kubelet/userns/userns_manager.go
+++ pkg/kubelet/userns/userns_manager.go
@@ -383,7 +383,7 @@ func (m *UsernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err err
 func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod, runtimeHandler string) (*runtimeapi.UserNamespace, error) {
 	featureEnabled := utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport)
 
-	if pod == nil || pod.Spec.HostUsers == nil || *pod.Spec.HostUsers {
+	if pod == nil || pod.Spec.HostUsers == nil {
 		// if the feature is enabled, specify to use the node mode...
 		if featureEnabled {
 			return &runtimeapi.UserNamespace{
@@ -393,15 +393,23 @@ func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod, runtimeHan
 		// ...otherwise don't even specify it
 		return nil, nil
 	}
+	// pod.Spec.HostUsers is set to true/false
+	if !featureEnabled {
+		return nil, fmt.Errorf("the feature gate %q is disabled: can't set spec.HostUsers", features.UserNamespacesSupport)
+	}
+	if *pod.Spec.HostUsers {
+		return &runtimeapi.UserNamespace{
+			Mode: runtimeapi.NamespaceMode_NODE,
+		}, nil
+	}
 
-	// if the pod requested a user namespace and the runtime doesn't support user namespaces then return an error,
-	// even when the feature gate is disabled.
+	// From here onwards, hostUsers=false and the feature gate is enabled.
+
+	// if the pod requested a user namespace and the runtime doesn't support user namespaces then return an error.
 	if handlerSupportsUserns, err := m.kl.HandlerSupportsUserNamespaces(runtimeHandler); err != nil {
 		return nil, err
 	} else if !handlerSupportsUserns {
-		return nil, fmt.Errorf("the handler %q does not support user namespaces", runtimeHandler)
-	} else if !featureEnabled {
-		return nil, nil
+		return nil, fmt.Errorf("the RuntimeClass handler %q does not support user namespaces", runtimeHandler)
 	}
 
 	m.lock.Lock()
```

Reply:
This is fixed now